src/lib/SaferNames/Name.hs

```diff
@@ -0,0 +1,478 @@
+  Abs (..), FreshAbs (..), Nest (..), NestPair (..),PlainBinderList, envMapWithKey,
+  AnnBinderP (..), AnnBinderListP (..), EnvE (..), RecEnv (..), RecEnvFrag (..),
+  AlphaEq (..), UnitE (..), VoidE, EmptyNest, PairE (..), MaybeE (..), ListE (..),
+  EitherE (..), LiftE (..), emptyFreshExt, composeFreshExt, idRenamer,
+  EqE, EqB, FreshBinder (..), lookupNameTraversal, extendScope) where
+
+import Control.Monad.Identity
+import Control.Monad.Writer.Strict
+import qualified Data.Set        as S
+import Unsafe.Coerce
+import Data.Text.Prettyprint.Doc  hiding (nest)
+import GHC.Exts (Constraint)
+
+import qualified SaferNames.LazyMap as LM
+
+import qualified Env as D
+
+data S = (:=>:) S S
+       | UnsafeMakeS
+       | UnitScope
+       | VoidScope
+
+-- TODO: we reuse the old `Name` to make use of the GlobalName name space while
+-- we're using both the old and new systems together.
+-- TODO: something like this instead:
+--    type Tag = T.Text
+--    data RawName = RawName Tag Int deriving (Show, Eq, Ord)
+type RawName = D.Name
+
+-- invariant: the raw name in `Name s` is contained in the list in the `ScopeVal s`
+newtype Name (n::S) = UnsafeMakeName RawName
+                      deriving (Show, Eq, Ord)
+
+newtype Env (n::S) (a:: *) = UnsafeMakeEnv (LM.LazyMap RawName a)
+                             deriving (Functor, Foldable, Traversable)
+type Scope n = Env n ()
+
+emptyEnv :: Env (n:=>:n) a
+emptyEnv = UnsafeMakeEnv mempty
+
+voidScopeEnv :: Env VoidScope a
+voidScopeEnv = UnsafeMakeEnv mempty
+
+infixl 1 <.>
+(<.>) :: Env (n1:=>:n2) a -> Env (n2:=>:n3) a -> Env (n1:=>:n3) a
+(<.>) (UnsafeMakeEnv m1) (UnsafeMakeEnv m2) = UnsafeMakeEnv (m2 <> m1)
+
+infixl 1 <>>
+(<>>) :: Env n a -> Env (n:=>:l) a -> Env l a
+(<>>) (UnsafeMakeEnv m1) (UnsafeMakeEnv m2) = UnsafeMakeEnv (m2 <> m1)
+
+envLookup :: Env n a -> Name n -> a
+envLookup (UnsafeMakeEnv m) (UnsafeMakeName name) =
+  case LM.lookup name m of
+    Just x -> x
+    Nothing -> error "Env lookup should never fail"
+
+envAsScope :: Env n a -> Scope n
+envAsScope (UnsafeMakeEnv m) = UnsafeMakeEnv $ LM.asUnitLazyMap m
+
+projectName :: Scope (n:=>:l) -> Name l -> Either (Name n) (Name (n:=>:l))
+projectName (UnsafeMakeEnv m) (UnsafeMakeName name) =
+  case LM.lookup name m of
+    Nothing -> Left  $ UnsafeMakeName name
+    Just () -> Right $ UnsafeMakeName name
+
+injectNameL :: FreshExt n l -> Name n -> Name l
+injectNameL _ (UnsafeMakeName name) = UnsafeMakeName name
+
+injectNameR :: Name (n:=>:l) -> Name l
+injectNameR (UnsafeMakeName name) = UnsafeMakeName name
+
+data PlainBinder (n::S) (l::S) where
+  UnsafeMakeBinder :: RawName -> PlainBinder n l
+  Ignore           :: PlainBinder n n
+
+type E = S -> *       -- expression-y things, covariant in the S param
+type B = S -> S -> *  -- binder-y things, covariant in the first param and
+                      -- contravariant in the second
+
+-- A `FreshExt n l` means that `l` extends `n` with only fresh names
```

Comment:
The scopes represent a list of possibly-repeated raw names (though actually we represent them with sets because that's all we need -- maybe we should change that). A fresh `FreshExt n l` means that (1) the `Scope n` is a prefix of the `Scope l` and (2) The names in the diff (the list `Scope l` once you take off the common prefix) do not appear in that common prefix. Is that a clearer way to put it?